File: bookshare-server/src/main/java/com/dmgburg/bookshareserver/MailingService.java
package com.dmgburg.bookshareserver;

import com.google.common.base.Charsets;
import com.google.common.io.Resources;
import org.apache.tomcat.util.http.fileupload.IOUtils;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.mail.Authenticator;
import javax.mail.Message;
import javax.mail.Multipart;
import javax.mail.PasswordAuthentication;
import javax.mail.Session;
import javax.mail.Transport;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeBodyPart;
import javax.mail.internet.MimeMessage;
import javax.mail.internet.MimeMultipart;
import java.io.IOException;
import java.net.URL;
import java.util.Properties;

@Component
public class MailingService {
    private final Session session;
    private final String baseURL;
    public MailingService(@Value("${dmgburg.mail.baseUrl}") String baseUrl,
                            @Value("${dmgburg.mail.username}") String username,
                            @Value("${dmgburg.mail.password}") String password) {
        Properties prop = new Properties();
        prop.setProperty("mail.smtp.auth", "true");
        prop.setProperty("mail.smtp.starttls.enable", "true");
        prop.setProperty("mail.smtp.host", "smtp.gmail.com");
        prop.setProperty("mail.smtp.port", "587");
        this.session = Session.getInstance(prop, new Authenticator() {
            @Override
            protected PasswordAuthentication getPasswordAuthentication() {
                return new PasswordAuthentication(username, password);
            }
        });
        this.baseURL = baseUrl;
    }

    public void sendMessage(String msg, String to) {
        try {
            Message message = new MimeMessage(session);
            message.setFrom(new InternetAddress("dmgburg@gmail.com"));
            message.setRecipients(
                    Message.RecipientType.TO, InternetAddress.parse("transformator00@mail.ru"));
            message.setSubject("Mail Subject");

            MimeBodyPart mimeBodyPart = new MimeBodyPart();
            mimeBodyPart.setContent(msg, "text/html");

            Multipart multipart = new MimeMultipart();
            multipart.addBodyPart(mimeBodyPart);

            message.setContent(multipart);

            Transport.send(message);
        } catch (Exception e){
            throw new RuntimeException(e);
        }
    }

    public void sendBookReqest(String holder, String requestor, String name, Long id) throws IOException {
        URL url = Resources.getResource("foo.txt");
        String text = Resources.toString(url, Charsets.UTF_8);
        text = text.replace("${requestor}", requestor);
        text = text.replace("${bookName}", name);
        text = text.replace("${baseURL}", baseURL);
        text = text.replace("${requestId}", baseURL);
        sendMessage(text, holder);
    }
}

File: bookshare-server/src/main/java/com/dmgburg/book/mail/MailingService.kt
package com.dmgburg.book.mail

import com.google.common.base.Charsets
import com.google.common.io.Resources
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.web.util.UriUtils
import java.util.*
import javax.mail.*
import javax.mail.internet.InternetAddress
import javax.mail.internet.MimeBodyPart
import javax.mail.internet.MimeMessage
import javax.mail.internet.MimeMultipart

interface MailingService {
    fun sendMessage(msg: String, subj: String, to: String)
    fun sendEmailConfirmation(to: String, hash: String)
}

@Component
class MailingServiceImpl(@Value("\${dmgburg.mail.baseUrl}") val baseUrl: String,
                         @Value("\${dmgburg.mail.host}") val serverHost: String,
                         @Value("\${dmgburg.mail.port}") val serverPort: Int,
                         @Value("\${dmgburg.mail.username}") val username: String,
                         @Value("\${dmgburg.mail.password}") val password: String?) : MailingService {
    private val session: Session
    private val baseURL: String
    override fun sendMessage(msg: String, subj: String, to: String) {
        try {
            val message: Message = MimeMessage(session)
            message.setFrom(InternetAddress(" support@co-books.com"))
            message.setRecipients(
                    Message.RecipientType.TO, InternetAddress.parse(to))
            message.subject = subj
            val mimeBodyPart = MimeBodyPart()
            mimeBodyPart.setContent(msg, "text/html")
            val multipart: Multipart = MimeMultipart()
            multipart.addBodyPart(mimeBodyPart)
            message.setContent(multipart)
            Transport.send(message)
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    override fun sendEmailConfirmation(to: String, hash: String) {
        val encoded = UriUtils.encode(to,"UTF-8")
        var text = "Please <a href=\"https://$baseUrl/api/user/public/confirmUser/${encoded}/$hash\">confirm</a> your email at co-books.com"
        sendMessage(text, "Please confirm your account at co-boocs.com", to)
    }

    fun sendBookReqest(holder: String, requestor: String, name: String, id: Long) {
        val url = Resources.getResource("foo.txt")
        var text = Resources.toString(url, Charsets.UTF_8)
        text = text.replace("\${requestor}", requestor)
        text = text.replace("\${bookName}", name)
        text = text.replace("\${baseURL}", baseURL)
        text = text.replace("\${requestId}", baseURL)
        sendMessage(text, "", holder)
    }

    init {
        val prop = Properties()
        prop.setProperty("mail.smtp.auth", "true")
        prop.setProperty("mail.smtp.host", serverHost)
        prop.setProperty("mail.smtp.port", serverPort.toString())
        prop.put("mail.transport.protocol", "smtp");
        prop.put("mail.smtp.socketFactory.class", "javax.net.ssl.SSLSocketFactory");
        session = Session.getInstance(prop, object : Authenticator() {
            override fun getPasswordAuthentication(): PasswordAuthentication {
                return PasswordAuthentication(username, password)
            }
        })
        baseURL = baseUrl
    }
}